pxd_fastpath.c

```diff
@@ -0,0 +1,940 @@
+	} else {
+		fp->bg_flush_enabled = true; // introduces high latency
+	}
+	fp->n_flush_wrsegs = MAX_WRITESEGS_FOR_FLUSH;
+
+	// congestion init
+	init_waitqueue_head(&fp->congestion_wait);
+	init_waitqueue_head(&fp->sync_event);
+	spin_lock_init(&fp->sync_lock);
+
+	atomic_set(&fp->nsync_active, 0);
+	atomic_set(&fp->nsync, 0);
+	atomic_set(&fp->nio_discard, 0);
+	atomic_set(&fp->nio_flush, 0);
+	atomic_set(&fp->nio_flush_nop, 0);
+	atomic_set(&fp->nio_preflush, 0);
+	atomic_set(&fp->nio_fua, 0);
+	atomic_set(&fp->nio_write, 0);
+	atomic_set(&fp->ncount,0);
+	atomic_set(&fp->nswitch,0);
+	atomic_set(&fp->nslowPath,0);
+	atomic_set(&fp->ncomplete,0);
+	atomic_set(&fp->nwrite_counter,0);
+
+	fp->tc = kzalloc(MAX_THREADS * sizeof(struct thread_context), GFP_NOIO);
+	if (!fp->tc) {
+		printk(KERN_ERR"Initializing backing volumes for pxd failed %d\n", err);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < nr_node_ids; i++) {
+		atomic_set(&fp->index[i], 0);
+	}
+
+	for (i = 0; i < MAX_THREADS; i++) {
+		struct thread_context *tc = &fp->tc[i];
+		tc->pxd_dev = pxd_dev;
+		spin_lock_init(&tc->lock);
+		init_waitqueue_head(&tc->pxd_event);
+		tc->pxd_thread = kthread_create_on_node(pxd_io_thread, tc, cpu_to_node(i),
+				"pxd%d:%llu", i, pxd_dev->dev_id);
+		if (IS_ERR(tc->pxd_thread)) {
+			pxd_printk("Init kthread for device %llu failed %lu\n",
+				pxd_dev->dev_id, PTR_ERR(tc->pxd_thread));
+			err = -EINVAL;
+			goto fail;
+		}
+
+		//
+		// NOTE this has to change for small sized, small queuedepth sync io.
+		// ibm mq issue. Will come in separate PR
+		//
+		kthread_bind(tc->pxd_thread, i);
+		set_user_nice(tc->pxd_thread, MIN_NICE);
+		wake_up_process(tc->pxd_thread);
+	}
+
+	enableFastPath(pxd_dev, true);
+
+	return 0;
+fail:
+	for (i = 0; i < MAX_THREADS; i++) {
+		struct thread_context *tc = &fp->tc[i];
+		if (tc->pxd_thread) kthread_stop(tc->pxd_thread);
+	}
+
+	if (fp->tc) kfree(fp->tc);
+	return err;
+}
+
+void pxd_fastpath_cleanup(struct pxd_device *pxd_dev) {
+	disableFastPath(pxd_dev);
+}
+
+/* fast path make request function, io entry point */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,4,0)
+blk_qc_t pxd_make_request_fastpath(struct request_queue *q, struct bio *bio)
+#else
+void pxd_make_request_fastpath(struct request_queue *q, struct bio *bio)
+#endif
+{
+	struct pxd_device *pxd_dev = q->queuedata;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
+	unsigned int rw = bio_op(bio);
+#else
+	unsigned int rw = bio_rw(bio);
+#endif
+	int cpu = smp_processor_id();
+	int thread = cpu % MAX_THREADS;
+
+	struct thread_context *tc;
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
+	if (!pxd_dev) {
+#else
+	if (rw == READA) rw = READ;
+	if (!pxd_dev || (rw != READ && rw != WRITE)) {
+#endif
+		printk(KERN_ERR"pxd basic sanity fail, pxd_device %p (%llu), rw %#x\n",
+				pxd_dev, (pxd_dev? pxd_dev->dev_id: (uint64_t)0), rw);
+		bio_io_error(bio);
+		return BLK_QC_RETVAL;
+	}
+
+	if (!pxd_dev->connected) {
```

Comment:
What's the case where fuse connection is up but dev connection is down?

Reply:
pxd_dev->connected follows fuse connected flag. All device carries it, and accesses it without any locking. While in pxd_context it was accessed with lock. This is now checked with every IO, so moved it within context to allow lock free access.

Reply:
But pxd_dev->connected also needs to be checked under lock rt?

Reply:
the thought process was its okay to check this outside lock, as it would probably process few more IO when it should have not allowed them, or fail IO when px is just starting up. The same could be extended to pxd_context->connected as well. I also see the logic around this has changed in master. Will resolve this when ready to merge.